File: xconf-angular-admin/src/main/java/com/comcast/xconf/admin/controller/change/ChangeController.java
/*
 * If not stated otherwise in this file or this component's LICENSE file the
 * following copyright and licenses apply:
 *
 * Copyright 2018 RDK Management
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.comcast.xconf.admin.controller.change;

import com.comcast.xconf.admin.service.change.ApprovedChangeCrudService;
import com.comcast.xconf.admin.service.change.ChangeCrudService;
import com.comcast.xconf.admin.service.change.TelemetryProfileChangeService;
import com.comcast.xconf.change.ApprovedChange;
import com.comcast.xconf.change.Change;
import com.comcast.xconf.shared.utils.ChangeUtils;
import com.comcast.xconf.shared.utils.PageUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Collections;
import java.util.List;
import java.util.Map;

import static com.comcast.xconf.admin.controller.change.ChangeController.URL;
import static org.springframework.http.HttpStatus.OK;

@RestController
@RequestMapping(URL)
public class ChangeController {

    public static final String URL = "/api/change";

    @Autowired
    private ChangeCrudService changeCrudService;

    @Autowired
    private ApprovedChangeCrudService approveChangeCrudService;

    @Autowired
    private TelemetryProfileChangeService telemetryProfileChangeService;

    @RequestMapping(method = RequestMethod.GET, value = "/approved")
    public ResponseEntity approved() {
        return new ResponseEntity<>(approveChangeCrudService.getAll(), OK);
    }

    @RequestMapping(method = RequestMethod.GET, value = "/approve/{changeId}")
    public ResponseEntity approve(@PathVariable String changeId) {
        telemetryProfileChangeService.approve(changeId);
        return new ResponseEntity<>(createHeadersWithEntitySize(), OK);
    }

    @RequestMapping(method = RequestMethod.GET, value = "/revert/{approveId}")
    public ResponseEntity revert(@PathVariable String approveId) {
        telemetryProfileChangeService.revert(approveId);
        return new ResponseEntity<>(createHeadersWithEntitySize(), OK);
    }

    @RequestMapping(method = RequestMethod.GET, value = "/cancel/{changeId}")
    public ResponseEntity cancel(@PathVariable String changeId) {
        telemetryProfileChangeService.cancel(changeId);
        return new ResponseEntity<>(createHeadersWithEntitySize(), OK);
    }

    @RequestMapping(method = RequestMethod.GET, value = "/changes/grouped/byId")
    public ResponseEntity getGroupedChanges(@RequestParam Integer pageNumber, @RequestParam Integer pageSize) {
        List<Change> changesPerPage = PageUtils.getPage(changeCrudService.getAll(), pageNumber, pageSize, ChangeUtils.ascByDateComparator());
        return new ResponseEntity<>(changeCrudService.groupChanges(changesPerPage), createHeadersWithEntitySize(), OK);
    }

    @RequestMapping(method = RequestMethod.GET, value = "/approved/grouped/byId")
    public ResponseEntity getGroupedApprovedChanges(@RequestParam Integer pageNumber, @RequestParam Integer pageSize) {
        List<ApprovedChange> changesPerPage = PageUtils.getPage(approveChangeCrudService.getAll(), pageNumber, pageSize, ChangeUtils.ascByDateComparator());
        return new ResponseEntity<>(Collections.singletonMap("changesPerPage", changesPerPage), createHeadersWithEntitySize(), OK);
    }

    @RequestMapping(method = RequestMethod.GET, value = "/entityIds")
    public ResponseEntity getChangedEntityIds() {
        return new ResponseEntity<>(changeCrudService.getChangedEntityIds(), OK);
    }

    @RequestMapping(method = RequestMethod.POST, value = "/approveChanges")
    public ResponseEntity approveChanges(@RequestBody List<String> changeIds) {
        return new ResponseEntity<>(telemetryProfileChangeService.approveChanges(changeIds), createHeadersWithEntitySize(), OK);
    }

    @RequestMapping(method = RequestMethod.POST, value = "/revertChanges")
    public ResponseEntity revertChanges(@RequestBody List<String> changeIds) {
        return new ResponseEntity<>(telemetryProfileChangeService.revertChanges(changeIds), OK);
    }

    @RequestMapping(value = "/approved/filtered", method = RequestMethod.POST)
    public ResponseEntity getApprovedFiltered(@RequestBody(required = false) Map<String, String> searchContext,
                                              @RequestParam(required = false, defaultValue = "1") Integer pageNumber,
                                              @RequestParam(required = false, defaultValue = "50") Integer pageSize) {
        List<Change> changes = changeCrudService.findByContext(searchContext);
        List<ApprovedChange> approvedChanges = approveChangeCrudService.findByContext(searchContext);
        List<ApprovedChange> approvedChangesPerPage = PageUtils.getPage(approvedChanges, pageNumber, pageSize, ChangeUtils.ascByDateComparator());
        return new ResponseEntity<>(Collections.singletonMap("changesPerPage", approvedChangesPerPage), createHeadersWithEntitySize(changes.size(), approvedChanges.size()), OK);
    }

    @RequestMapping(value = "/changes/filtered", method = RequestMethod.POST)
    public ResponseEntity getChangesFiltered(@RequestBody(required = false) Map<String, String> searchContext,
                                             @RequestParam(required = false, defaultValue = "1") Integer pageNumber,
                                             @RequestParam(required = false, defaultValue = "50") Integer pageSize) {
        List<Change> changes = changeCrudService.findByContext(searchContext);
        List<Change> changesPerPage = PageUtils.getPage(changes, pageNumber, pageSize, ChangeUtils.ascByDateComparator());
        List<ApprovedChange> approvedChanges = approveChangeCrudService.findByContext(searchContext);
        return new ResponseEntity<>(changeCrudService.groupChanges(changesPerPage), createHeadersWithEntitySize(changes.size(), approvedChanges.size()), OK);
    }

    private HttpHeaders createHeadersWithEntitySize() {
        return createHeadersWithEntitySize(
                changeCrudService.getAll().size(),
                approveChangeCrudService.getAll().size()
        );
    }

    private HttpHeaders createHeadersWithEntitySize(int pendingChangesSize, int approvedChangesSize) {
        HttpHeaders headers = new HttpHeaders();
        headers.add("pendingChangesSize", String.valueOf(pendingChangesSize));
        headers.add("approvedChangesSize", String.valueOf(approvedChangesSize));
        return headers;
    }
}

File: xconf-angular-admin/src/main/webapp/app/xconf/changes/change.controller.js
/*
 * If not stated otherwise in this file or this component's LICENSE file the
 * following copyright and licenses apply:
 *
 * Copyright 2018 RDK Management
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

(function() {
    'use strict';

    angular
        .module('app.change')
        .controller('ChangeController', controller);

    controller.$inject=['$scope', '$rootScope', 'changeService', 'alertsService', 'CHANGE_TYPE', 'utilsService', 'dialogs', 'paginationService', 'CHANGE_OPERATION', 'ENTITY_TYPE', '$controller', 'authUtilsService'];
    function controller($scope, $rootScope, changeService, alertsService, CHANGE_TYPE, utilsService, dialogs, paginationService, CHANGE_OPERATION, ENTITY_TYPE, $controller, authUtils) {
        var vm = this;

        angular.extend(vm, $controller('MainController', {
            $scope: $scope
        }));

        vm.currentChangeType = CHANGE_TYPE.PENDING;
        vm.CHANGE_TYPE = CHANGE_TYPE;
        vm.ENTITY_TYPE = ENTITY_TYPE;
        vm.changes = [];
        vm.changeDiffs = {};
        vm.approvedChangesSize = null;
        vm.pendingChangesSize = null;
        vm.paginationStorageKey = 'changesPageSize';
        vm.availablePageSizes = paginationService.getAvailablePageSizes();
        vm.pageSize = paginationService.getPageSize(vm.paginationStorageKey);
        vm.pageSizeModel = vm.pageSize;
        vm.pageNumber = paginationService.getPageNumber();
        vm.numPages = 0;
        vm.changesForMultipleOperation = [];
        vm.CHANGE_OPERATION = CHANGE_OPERATION;

        vm.isNullOrUndefinedOrEmptyObject = utilsService.isNullOrUndefinedOrEmptyObject;
        vm.errorMessageById = {};

        vm.searchOptions = {
            data: [
                {
                    "name": {
                        friendlyName: "Entity",
                        apiArgs: ['ENTITY']
                    }
                },
                {
                    "name": {
                        friendlyName: "User",
                        apiArgs: ["AUTHOR"]
                    }
                }
            ]
        };
        vm.cancel = cancel;
        vm.getChanges = getChanges;
        vm.getApprovedChanges = getApprovedChanges;
        vm.getEntityView = changeService.getEntityView;
        vm.getEntityName = changeService.getEntityName;
        vm.getSizeByType = getSizeByType;
        vm.getChangesByType = getChangesByType;
        vm.changePageSize = changePageSize;
        vm.updateChangeList = updateChangeList;
        vm.applySelectedChanges = applySelectedChanges;
        vm.isAddedToMultipleOperation = isAddedToMultipleOperation;
        vm.isEmptyString = utilsService.isEmptyString;
        vm.canWriteChangeAndTelemetry = canWriteChangeAndTelemetry;

        init();

        function init() {
            getChanges();
        }

        $scope.$on('$locationChangeSuccess', function () {
            if (paginationService.paginationSettingsInLocationHaveChanged(vm.pageNumber, vm.pageSize)) {
                vm.pageSize = paginationService.getPageSize(vm.paginationStorageKey);
                vm.pageSizeModel = vm.pageSize;
                paginationService.saveDefaultPageSize(vm.pageSize, vm.paginationStorageKey);
                vm.pageNumber = paginationService.getPageNumber();
                init();
            }
        });

        $scope.$on('search-entities', function(event, data) {
            vm.searchParam = data.searchParam;
            var changePromise;
            if (vm.currentChangeType === CHANGE_TYPE.APPROVED) {
                changePromise = changeService.getFilteredApprovedChanges(vm.pageSize, vm.pageNumber, vm.searchParam);
            } else {
                changePromise = changeService.getFilteredChanges(vm.pageSize, vm.pageNumber, vm.searchParam);
            }
            changePromise.then(responseHandler, alertsService.errorHandler);
        });

        function getApprovedChanges() {
            cleanUpChangeIds(vm.currentChangeType, CHANGE_TYPE.APPROVED, vm.changesForMultipleOperation);
            vm.currentChangeType = CHANGE_TYPE.APPROVED;
            changeService.getFilteredApprovedChanges(vm.pageSize, vm.pageNumber, vm.searchParam).then(
                responseHandler, alertsService.errorHandler
            );
        }

        function responseHandler(resp) {
            vm.changes = resp.data;
            getSizes(resp);
        }

        function getChanges() {
            cleanUpChangeIds(vm.currentChangeType, CHANGE_TYPE.PENDING, vm.changesForMultipleOperation);
            vm.currentChangeType = CHANGE_TYPE.PENDING;
            changeService.getFilteredChanges(vm.pageSize, vm.pageNumber, vm.searchParam).then(
                responseHandler, alertsService.errorHandler
            );
        }

        function cancel(change) {
            var dlg = dialogs.confirm('Cancel confirmation', '<span class="break-word-inline">Are you sure you want to cancel change of ' + vm.getEntityName(change) + " entity? </span>");
            dlg.result.then(function(btn) {
                changeService.cancel(change.id).then(function (resp) {
                    alertsService.showSuccessMessage({message: 'Change of ' + vm.getEntityName(change) + ' entity successfully canceled'});
                    utilsService.removeItemFromListById(vm.changes, change.id);
                    utilsService.removeItemFromListById(vm.changesForMultipleOperation, change.id);
                    getSizes(resp);
                    vm.pageNumber = getPageNumberAfterUpdate(vm.currentChangeType, vm.pageSize, vm.pageNumber);
                    getChangesByType(vm.currentChangeType);
                }, alertsService.errorHandler);
            }, function() {
                // click cancel
            });
        }

        function getSizes(resp) {
            vm.approvedChangesSize = resp.headers('approvedChangesSize');
            vm.pendingChangesSize = resp.headers('pendingChangesSize');
        }

        function getSizeByType(type) {
            return type === CHANGE_TYPE.PENDING ? vm.pendingChangesSize : vm.approvedChangesSize;
        }

        function getChangesByType(type) {
            if (type === CHANGE_TYPE.PENDING) {
                getChanges();
            } else {
                getApprovedChanges();
            }
        }

        function changePageSize(pageSizeModel) {
            vm.pageSize = pageSizeModel;
            paginationService.saveDefaultPageSize(vm.pageSize, vm.paginationStorageKey);
            getChangesByType(vm.currentChangeType);
        }

        function getPageNumberAfterUpdate(currentPendingChangeType, pageSize, pageNumber) {
            var numberOfPagesAfterDeletion = Math.ceil((getSizeByType(currentPendingChangeType)) / pageSize);
            var newPageNumber = pageNumber > numberOfPagesAfterDeletion && numberOfPagesAfterDeletion !== 0 ? numberOfPagesAfterDeletion : pageNumber;
            return newPageNumber;
        }

        function updateChangeList(checked, change) {
           if (checked) {
               vm.changesForMultipleOperation.push(change);
           } else {
               utilsService.removeItemFromArrayWithDeepEquals(vm.changesForMultipleOperation, change);
           }
           if (!vm.isEmptyString(vm.errorMessageById[change.id])) {
               vm.errorMessageById[change.id] = '';
           }
        }

        function applySelectedChanges(operation) {
            var dlg = dialogs.confirm('Approve confirmation', '<span class="break-word-inline">Are you sure you want to ' + operation.toLowerCase() + ' ' + vm.changesForMultipleOperation.length + ' selected changes?</span>');
            dlg.result.then(function(btn) {
                applyChanges(vm.changesForMultipleOperation);
            }, function() {
                // click cancel
            });
        }

        function applyChanges(changes) {
            var changeIds = _.map(changes, function(change){return change.id});
            var changedPromise;
            if (vm.currentChangeType === vm.CHANGE_TYPE.PENDING) {
                changedPromise = changeService.approveChanges(changeIds);
            } else {
                changedPromise = changeService.revertChanges(changeIds);
            }
            changedPromise.then(function (resp) {
                vm.errorMessageById = resp.data;
                if (utilsService.isMapEmpty(vm.errorMessageById)) {
                    showSuccessApproveRevertMessage();
                } else {
                    alertsService.showError({title: 'Errors occurred when trying to revert selected changes'});
                }
                getChangesByType(vm.currentChangeType);
                cleanUpSelectedChanges(vm.changesForMultipleOperation, _.keys(vm.errorMessageById));
            }, function (error) {
                alertsService.showError({title: 'Error', message: error.data.message});
                getChangesByType(vm.currentChangeType);
            });
        }

        function showSuccessApproveRevertMessage() {
            if (vm.currentChangeType === vm.CHANGE_TYPE.PENDING) {
                alertsService.showSuccessMessage({message: 'Selected changes have been successfully approved and unselected changes canceled'});
            } else {
                alertsService.showSuccessMessage({message: 'Selected changes were reverted successfully'});
            }
        }

        function cleanUpSelectedChanges(selectedChanges, changeIdsWithError) {
            console.log(changeIdsWithError);
            _.each(selectedChanges, function(change) {
                if (!changeIdsWithError.includes(change.id)) {
                    utilsService.removeItemFromListById(selectedChanges, change.id);
                }
            });
        }

        function isAddedToMultipleOperation(change) {
            for(var i = 0; i < vm.changesForMultipleOperation.length; i++) {
                if (angular.equals(vm.changesForMultipleOperation[i], change)) {
                    return true
                }
            }
        }

        function cleanUpChangeIds(currentPendingChangeType, newPendingChangeType, changesForMultipleOperation) {
            if (currentPendingChangeType !== newPendingChangeType) {
                changesForMultipleOperation.length = 0;
                clearErrorMessages();
            }
        }

        function clearErrorMessages() {
            vm.errorMessageById = {};
        }

        function canWriteChangeAndTelemetry() {
            return authUtils.canWriteChangesByApplication($rootScope.applicationType) && authUtils.canWriteTelemetryByApplication($rootScope.applicationType);
        }
    }
})();